### PR TITLE
Move several of the platform specific threading mechanisms to C++11 standard.

### PR DESCRIPTION
**Pull request by jimfcarroll**
Eliminated several of the custom threading mechanisms that are now supported by the C++ standard. This allows us to remove several of the platform specific files.

## Motivation and Context
Less code to maintain going forward.

## How Has This Been Tested?
Manually. Plus all of the tests pass.

## Types of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

File: xbmc/threads/Condition.h [hunk 1]
```diff
@@ -20,14 +20,59 @@
...
+#include <condition_variable>
+#include <chrono>
 
 namespace XbmcThreads
 {
 
+  /**
+   * This is a thin wrapper around boost::condition_variable. It is subject
+   *  to "spurious returns" as it is built on boost which is built on posix
+   *  on many of our platforms.
```
Comment (wsnipex) [inline xbmc/threads/Condition.h]: comment is probably not valid anymore?

File: xbmc/threads/platform/win/CMakeLists.txt [hunk 2]
```diff
@@ -1,7 +1,6 @@
 set(HEADERS Condition.h
```
Comment (Rechi) [inline xbmc/threads/platform/win/CMakeLists.txt]: file got removed

File: xbmc/threads/Condition.h (hunk 1 above)
Comment (pkerling) [reply xbmc/threads/Condition.h]: It wasn't even valid before, boost is long gone :-)

File: xbmc/threads/test/TestThreadLocal.cpp [hunk 3]
```diff
@@ -35,7 +33,7 @@ class Thingy
 
 Thingy* staticThingy = NULL;
 CEvent gate;
-ThreadLocal<Thingy> staticThreadLocal;
+thread_local Thingy* staticThreadLocal;
```
Comment (pkerling) [inline xbmc/threads/test/TestThreadLocal.cpp]: What's the point of this test if we use `thread_local` anyway?

File: xbmc/threads/Condition.h [hunk 4]
```diff
@@ -20,14 +20,59 @@
...
+   */
+  class ConditionVariable : public NonCopyable
```
Comment (pkerling) [inline xbmc/threads/Condition.h]: Do we want to use `NonCopyable` in new code? Wouldn't it be better to use `= delete`?

File: xbmc/threads/Condition.h (hunk 1 above)
Comment (MrMC) [reply xbmc/threads/Condition.h]: heh, "built on boost which is built on posix", conditions in posix can and will 'spurious return', you must check for it. The comment is quite valid but worded wrong now.

**Comment by MrMC** [comment]
Tested on iOS builds ? Last time I checked, thread local was still forbidden. 

Oh and hi jimfcarroll , davilla here :)

**Comment by Rechi** [comment]
`thread_local` compiles for iOS if `iphoneos-version-min` is bumped to 9, not runtime tested it.

**Comment by jimfcarroll** [comment]
Ah. Hey davilla. I thought thread_local was C++11 specific. Obviously something failed somewhere :-)

File: xbmc/threads/Condition.h (hunk 4 above)
Comment (jimfcarroll) [reply xbmc/threads/Condition.h]: @pkerling I can fix this. This isn't new code though. The class was just moved and modified from a platform specific version to make it non-platform specific.

File: xbmc/threads/test/TestThreadLocal.cpp (hunk 3 above)
Comment (jimfcarroll) [reply xbmc/threads/test/TestThreadLocal.cpp]: It also tests Event

File: xbmc/threads/Condition.h (hunk 1 above)
Comment (jimfcarroll) [reply xbmc/threads/Condition.h]: Yeah. It's c-n-p from a platform specific file. I can fix it.

**Comment by pkerling** [comment]
We already require C++11.

File: xbmc/threads/Condition.h (hunk 4 above)
Comment (pkerling) [reply xbmc/threads/Condition.h]: Ah OK, didn't check where this was from. Still, can as well fix it while we're at it :)

File: xbmc/threads/test/TestThreadLocal.cpp (hunk 3 above)
Comment (pkerling) [reply xbmc/threads/test/TestThreadLocal.cpp]: There is already `TestEvent.cpp` - this file was specifically for `ThreadLocal`

Comment (jimfcarroll) [reply xbmc/threads/test/TestThreadLocal.cpp]: I know. I wrote it about 6 years ago :-) It uses CEvent also so I left it. I can trim it down.

File: xbmc/threads/platform/win/CMakeLists.txt (hunk 2 above)
Comment (jimfcarroll) [reply xbmc/threads/platform/win/CMakeLists.txt]: Yup. Will fix. Thanks.

**Comment by jimfcarroll** [comment]
Exactly. So thread_local *should* work http://en.cppreference.com/w/cpp/keyword/thread_local

**Comment by MrMC** [comment]
thread_local is supported for iOS 9 and later, but not for iOS 8.4 or earlier.

**Comment by jimfcarroll** [comment]
@MrMC what's your recommendation? I can roll back the ThreadLocal removal or we can move to full C++11 support in iOS. Up to you.

**Comment by MrMC** [comment]
Not really up to me :) Depends on which iOS version is being targeted for Kodi. For MrMC fork, we target 10.x. 

Also should check android very carefully, their compiler/libc/libc++ implementations are known for being incomplete.

**Comment by jimfcarroll** [comment]
Well, it looks like the build succeeded on Android and we know why it failed on Windows (I forgot to remove a reference to a deleted header from the cmake file). So it's up to the team how they want to proceed on iOS. I have plenty of other things to address at this point so someone needs to just let me know.

File: xbmc/threads/test/TestThreadLocal.cpp (hunk 3 above)
Comment (pkerling) [reply xbmc/threads/test/TestThreadLocal.cpp]: Is there some specific interaction between `CEvent` and `thread_local`? If not, moving the tests to `TestEvent` would make them clearer about their purpose.

**Comment by Memphiz** [comment]
Looks like osx doesn’t like that either. Well even if slack told me that my tiny niche platforms are not supposed to stop progress - this change seems to need a bump in:

1. Deployment target for ios from 6.0 to 9.0 (so minmum ios version required will be 9.0 then)
2. the used xcode version on jenkins for ios from xcode7.2 to 9.0
3. The used ios sdk on jenkins from 9.2 to 10.11
4. The used xcode version on jenkins for osx from xcode6.1.1 to 9.0
5. the used osx sdk in jenkins from 10.10 to 10.12
6. All this adapted in the readme.ios and readme osx

If i got it right the deployment target for osx can stay on 10.8

Does this change warrent the risk involved with that heavy tool and sdk update? Is someone ready to recover the platforms from resulting fallout if there is any?

Sure i am all for updating tools but i am not really there for fixing broken stuff due to those changes. Maybe @rechi or @FernetMenta could do some fixes if needed for osx but i think i am the only one caring about ios atm.

@MartijnKaijser just for completeness - kodi runs on non jailbroken devices and it seems a lot of users don’t care about the involved resign proccess every 7 days (or a lot of them got dev certificates)

**Comment by hudokkow** [comment]
Don't use @Rechi on a commit name. Poor Rechi will be pinged for all eternity... 😉

**Comment by jimfcarroll** [comment]
Maybe. But he gave me the patch so I wanted him to have the credit (a.k.a. "blame") :-)

**Comment by MrMC** [comment]
git commit --amend --author="Rechi <Rechi@users.noreply.github.com>"

**Comment by popcornmix** [comment]
Builds and runs without obvious issues on Pi.

**Comment by jimfcarroll** [comment]
jenkins build this please

**Comment by jimfcarroll** [comment]
jenkins build this please

**Comment by jimfcarroll** [comment]
jenkins build this please

**Comment by MartijnKaijser** [comment]
Jenkins doesn't always like rebuilding.
Perhaps a squash and force push can get it going again

**Comment by jimfcarroll** [comment]
Yeah. I was hoping to do that after I got a clean build on all platforms but I'll give it a shot.

**Comment by jimfcarroll** [comment]
@FernetMenta @notspiff This builds (I built it on all platforms on Jenkins manually). I also just kicked off another build. Is it okay to commit it once Jenkins finishes?

**Comment by notspiff** [comment]
all good from my side.

**Comment by Rechi** [comment]
@jimfcarroll please wait for #13468 getting merged

**Comment by jimfcarroll** [comment]
@Rechi this has those changes already

**Comment by notspiff** [comment]
yes, which is the point. they should be brought in separately.

**Comment by Rechi** [comment]
the documentation fixes are missing here

**Comment by jimfcarroll** [comment]
Okay

File: xbmc/threads/Condition.h [hunk 5]
```diff
@@ -20,14 +20,59 @@
...
 
 namespace XbmcThreads
 {
 
+  /**
+   * This is a thin wrapper around std::condition_variable_any. It is subject
+   *  to "spurious returns" as it is built on boost which is built on posix
+   *  on many of our platforms.
+   */
+  class ConditionVariable : public NonCopyable
```
Comment (Paxxi) [inline xbmc/threads/Condition.h]: Can get rid of NonCopyable and just delete the copy constructors.
Not necessary but cleans things up a bit

**Review by Memphiz — APPROVED** [review]

File: xbmc/threads/platform/pthreads/Implementation.cpp [hunk 6]
```diff
@@ -1,60 +0,0 @@
...
-
-    static bool setRecursiveAttr() 
-    {
-      static bool alreadyCalled = false; // initialized to 0 in the data segment prior to startup init code running
-      if (!alreadyCalled)
-      {
-        pthread_mutexattr_init(&recursiveAttr);
-        pthread_mutexattr_settype(&recursiveAttr,PTHREAD_MUTEX_RECURSIVE);
-#if !defined(__arm__) && !defined(TARGET_ANDROID)
-        pthread_mutexattr_setprotocol(&recursiveAttr,PTHREAD_PRIO_INHERIT);
```
Comment (FernetMenta) [inline xbmc/threads/platform/pthreads/Implementation.cpp]: @jimfcarroll dropping PTHREAD_PRIO_INHERIT is no go. Why did you do this? I told you on slack that this is needed

**Pull request merged**